pymilvus/milvus_client/milvus_client.py

```diff
@@ -0,0 +1,801 @@
+        batch_size: int = 100,
+        partition: str = None,
+        progress_bar: bool = False,
+    ) -> List[Union[str, int]]:
+        """Insert data into the collection.
+
+        If the Milvus Client was initiated without an existing Collection, the first dict passed
+        in will be used to initiate the collection.
+
+        Args:
+            data (List[Dict[str, any]]): A list of dicts to pass in. If list not provided, will
+                cast to list.
+            timeout (int, optional): The timeout to use, will override init timeout. Defaults
+                to None.
+            batch_size (int, optional): The batch size to perform inputs with. Defaults to 100.
+            partition (str, optional): Which partition to insert into. Defaults to None.
+            progress_bar (bool, optional): Whether to display a progress bar for the input.
+                Defaults to False.
+
+        Raises:
+            DataNotMatchException: If the data has misssing fields an exception will be thrown.
+            MilvusException: General Milvus error on insert.
+
+        Returns:
+            List[Union[str, int]]: A list of primary keys that were inserted.
+        """
+        # If no data provided, we cannot input anything
+        if len(data) == 0:
+            return []
+
+        # If the collection hasnt been initialized, initialize it
+        with self.concurrent_lock:
+            if self.collection is None:
+                self._init(data[0])
+
+        # Dont include the primary key if auto_id is true and they included it in data
+        ignore_pk = self.pk_field if self.collection.schema.auto_id else None
+        insert_dict = {}
+        pks = []
+
+
+        for k in data:
+            for key, value in k.items():
+                if key in self.fields:
+                    insert_dict.setdefault(key, []).append(value)
+
+        # Insert the data in batches
+        for i in tqdm(range(0, len(data), batch_size), disable=not progress_bar):
+            # Convert dict to list of lists batch for insertion
+            try:
+                insert_batch = [insert_dict[key][i : i + batch_size] for key in self.fields if key != ignore_pk]
+            except KeyError as ex:
+                logger.error(
+                    "Malformed data, one of the inserts does not contain all fields required."
+                )
+                raise ex
+            # Insert into the collection.
+            try:
+                res = self.collection.insert(
+                    insert_batch,
+                    timeout=timeout or self.timeout,
+                    partition_name=partition,
+                )
+                pks.extend(res.primary_keys)
+            except MilvusException as ex:
+                logger.error(
+                    "Failed to insert batch starting at entity: %s/%s", str(i), str(len(data))
+                )
+                raise ex
+        return pks
+
+    def upsert_data(
```

Comment:
change to upsert implementation in the future.
comment it out the performance might be really bad due to extra delete

Reply:
agree, atomicity can't be promised if we implement the `upsert` on the client side.

Reply:
Yes, this is something that I want to replace when upsert is supported. For now most integrations and users need some kind of upsert and this is the route we are recommending, so this is just a temporary solution. Once 2.3 stable is released I will replace it. I will add a comment in docstring that emphasizes that this operation is slow and not guaranteed.